command-line/StatesAndEvents.py

```diff
@@ -0,0 +1,121 @@
+from enum import Enum
+
+from typing import List, Union
+
+from dataclasses import dataclass
+
+
+class VCUEvents (Enum):
+    """
+    The VCU Events, refer to VCU\Phantom\data_structures\vcu_common.h
+    """
+    EVENT_APPS1_RANGE_FAULT = 1
+    EVENT_APPS2_RANGE_FAULT = 2
+    EVENT_BSE_RANGE_FAULT = 3
+    EVENT_FP_DIFF_FAULT = 4
+    EVENT_RESET_CAR = 5
+    EVENT_READY_TO_DRIVE = 6
+    EVENT_TRACTIVE_ON = 7
+    EVENT_TRACTIVE_OFF = 8 
+    EVENT_BRAKE_PLAUSIBILITY_CLEARED = 9
+    EVENT_BRAKE_PLAUSIBILITY_FAULT = 10
+    EVENT_UNRESPONSIVE_APPS = 11
+class VCUStates(Enum):
+    """
+    The VCU states, refer to VCU\Phantom\data_structures\vcu_common.h
+    """
+    TRACTIVE_OFF = 0
+    TRACTIVE_ON = 1 
+    RUNNING = 2
+    MINOR_FAULT = 3
+    SEVERE_FAULT = 4
+
+@dataclass
+class EventData:
+    EVENT: VCUEvents
+    TIME: int
+
+@dataclass
+class StateData:
+    STATE: VCUStates
+    TIME: Union[int, None]
+
+class ResponseVCU:
+    """
+    The data structure in which the VCU returns values via UART communication
+    after sending a sequence of values
+
+    Allows for the same events to be stored if triggered at different times
+    It is necessary for the this class to specify the VCU state
+    
+    Stores an array of events and their time of trigger
+    Stores the state of the VCU and the time of change (if changed)
+    """
+
+    def __init__(self):
+        self._events: List[EventData] = []
+        self._state : StateData = None
+    
+    def add_event(self, event_name: VCUEvents, event_time: int):
+        """
+        Add an event to the list of events
+
+        :param event_name: a VCU event
+        :param event_time: the relative time of the event
+        """
+
+        #check that the event exists in the available VCU Events enum
+        if not isinstance(event_name, VCUEvents):
+            raise EventError(f"Invalid event type {event_name}, should be of type {VCUEvents.__name__}")
+        
+        event_data = EventData(EVENT=event_name,TIME=event_time)
+        
+        if not self._check_duplicate(event_data):
```

Comment:
If we had it as set, we would need to write our own method for hashing since classes are unhashable. Please refer to the below article
https://stackoverflow.com/questions/10547343/add-object-into-pythons-set-collection-and-determine-by-objects-attribute
We would also need to hash by two attributes (event and time), and including overrides of built in methods gets very messy

Moreover, we would need to cast self.events = set({}) or else it would act as a dictionnary
Overall, I think the list makes the code look more simple
It might later be useful to order the list by time of event too in case we want this use case

I'm not entirely sure what you mean by the VCU spam. The TIME attribute is in ms, and if there are two events in the exact same time, it will then raise the necessary error which would indicate high volume of events spam. We can always save the VCU returned string to debug a failed test too, but I don't think the data structure should allow us to add illegal sequences of events

Reply:
this is why dataclasses are so awesome, they handle all the boilerplate code for dunder (built-in) methods https://www.onlinegdb.com/zNjt8hrIY

regardless, I see what you mean for VCU spam. We'll just need to handle the exception gracefully in the application code. Resolved!